add github package registry (#13)
navarroaxel/navarroaxel.github.io: blog/publish-an-npm-to-github-packages.md

# Publish an npm to GitHub packages

Sometimes in website and backend projects we found common components (React, utils, validations, etc) and, if we follow the DRY concept, we should find a way to create a private package and install it in every project that needs it.

We're going to review how to build and publish JavaScript packages using npm.

## The GitHub solution

> GitHub provides the [GitHub Package Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry) to publish private npm packages. We can also use it for Docker images, and libraries for other languages like Ruby, but we're going to focus on the npm solution.

Reviewer: ```suggestion
GitHub provides the [GitHub Package Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry) to publish private npm packages. We can also use it for Docker images and libraries for other languages like Ruby, but we're going to focus on the npm solution.
```

### The publish configuration

You should add the `publishConfig` section in the `package.json` file to publish to the GitHub registry.

```json
"publishConfig": {
  "registry": "https://npm.pkg.github.com"
},
```

### The workflow

This workflow will publish a package to the GitHub Registry every time we create a release in the GitHub repository:

```yaml
name: Publish
on:
  release:
    types: [created]
jobs:
  publish:
    runs-on: ubuntu-latest
    permissions:
      contents: read
      packages: write
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: 16
      - run: npm install
      - run: npm run build
      - run: |
          echo @lexacode:https://npm.pkg.github.com/ > build/.npmrc
          echo '//npm.pkg.github.com/:_authToken=${NPM_TOKEN}' >> build/.npmrc
      - run: npm publish
        working-directory: ./build
        env:
          NPM_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

The [`permissions`](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#permissions) modifies the default permissions granted to the `GITHUB_TOKEN`.

> The important script in this workflow is create a `.npmrc` file inside the `build/` directory before publish to the registry.

Reviewer: ```suggestion
In this workflow is to create a `.npmrc` file inside the `build/` directory before publishing to the registry.
```

Author: ```suggestion
This workflow creates a `.npmrc` file inside the `build/` directory before publishing the package to the registry.
```

```bash
echo @lexacode:https://npm.pkg.github.com/ > build/.npmrc
echo '//npm.pkg.github.com/:_authToken=${NPM_TOKEN}' >> build/.npmrc
```

> 🧠 Remember that your organization name, e.g. `lexacode`, should be in `kebab-case`, **no mayus allowed**.

Reviewer: ```suggestion
🧠 Remember that your organization name, e.g. `lexacode`, should be in `kebab-case`, **no uppercase allowed**.
```
is `mayus` a word?
